File: readium/shared/src/main/java/org/readium/r2/shared/util/zip/compress/archivers/zip/AbstractUnicodeExtraField.java
/*
 * Copyright 2023 Readium Foundation. All rights reserved.
 * Use of this source code is governed by the BSD-style license
 * available in the top-level LICENSE file of the project.
 */

/*
 *  Licensed to the Apache Software Foundation (ASF) under one or more
 *  contributor license agreements.  See the NOTICE file distributed with
 *  this work for additional information regarding copyright ownership.
 *  The ASF licenses this file to You under the Apache License, Version 2.0
 *  (the "License"); you may not use this file except in compliance with
 *  the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */

package org.readium.r2.shared.util.zip.compress.archivers.zip;

import static java.nio.charset.StandardCharsets.UTF_8;

import java.util.zip.CRC32;
import java.util.zip.ZipException;

/**
 * A common base class for Unicode extra information extra fields.
 * @NotThreadSafe
 */
public abstract class AbstractUnicodeExtraField implements ZipExtraField {
    private long nameCRC32;
    private byte[] unicodeName;
    private byte[] data;

    protected AbstractUnicodeExtraField() {
    }

    /**
     * Assemble as unicode extension from the name/comment and
     * encoding of the original zip entry.
     *
     * @param text The file name or comment.
     * @param bytes The encoded of the file name or comment in the zip
     * file.
     */
    protected AbstractUnicodeExtraField(final String text, final byte[] bytes) {
        this(text, bytes, 0, bytes.length);
    }

    /**
     * Assemble as unicode extension from the name/comment and
     * encoding of the original zip entry.
     *
     * @param text The file name or comment.
     * @param bytes The encoded of the file name or comment in the zip
     * file.
     * @param off The offset of the encoded file name or comment in
     * {@code bytes}.
     * @param len The length of the encoded file name or comment in
     * {@code bytes}.
     */
    protected AbstractUnicodeExtraField(final String text, final byte[] bytes, final int off, final int len) {
        final CRC32 crc32 = new CRC32();
        crc32.update(bytes, off, len);
        nameCRC32 = crc32.getValue();

        unicodeName = text.getBytes(UTF_8);
    }

    private void assembleData() {
        if (unicodeName == null) {
            return;
        }

        data = new byte[5 + unicodeName.length];
        // version 1
        data[0] = 0x01;
        System.arraycopy(ZipLong.getBytes(nameCRC32), 0, data, 1, 4);
        System.arraycopy(unicodeName, 0, data, 5, unicodeName.length);
    }

    @Override
    public byte[] getCentralDirectoryData() {
        if (data == null) {
            this.assembleData();
        }
        byte[] b = null;
        if (data != null) {
            b = new byte[data.length];
            System.arraycopy(data, 0, b, 0, b.length);
        }
        return b;
    }

    @Override
    public ZipShort getCentralDirectoryLength() {
        if (data == null) {
            assembleData();
        }
        return new ZipShort(data != null ? data.length : 0);
    }

    @Override
    public byte[] getLocalFileDataData() {
        return getCentralDirectoryData();
    }

    @Override
    public ZipShort getLocalFileDataLength() {
        return getCentralDirectoryLength();
    }

    /**
     * @return The CRC32 checksum of the file name or comment as
     *         encoded in the central directory of the zip file.
     */
    public long getNameCRC32() {
        return nameCRC32;
    }

    /**
     * @return The UTF-8 encoded name.
     */
    public byte[] getUnicodeName() {
        byte[] b = null;
        if (unicodeName != null) {
            b = new byte[unicodeName.length];
            System.arraycopy(unicodeName, 0, b, 0, b.length);
        }
        return b;
    }

    /**
     * Doesn't do anything special since this class always uses the
     * same data in central directory and local file data.
     */
    @Override
    public void parseFromCentralDirectoryData(final byte[] buffer, final int offset,
                                              final int length)
        throws ZipException {
        parseFromLocalFileData(buffer, offset, length);
    }

    @Override
    public void parseFromLocalFileData(final byte[] buffer, final int offset, final int length)
        throws ZipException {

        if (length < 5) {
            throw new ZipException("UniCode path extra data must have at least 5 bytes.");
        }

        final int version = buffer[offset];

        if (version != 0x01) {
            throw new ZipException("Unsupported version [" + version
                                   + "] for UniCode path extra data.");
        }

        nameCRC32 = ZipLong.getValue(buffer, offset + 1);
        unicodeName = new byte[length - 5];
        System.arraycopy(buffer, offset + 5, unicodeName, 0, length - 5);
        data = null;
    }

    /**
     * @param nameCRC32 The CRC32 checksum of the file name as encoded
     *         in the central directory of the zip file to set.
     */
    public void setNameCRC32(final long nameCRC32) {
        this.nameCRC32 = nameCRC32;
        data = null;
    }

    /**
     * @param unicodeName The UTF-8 encoded name to set.
     */
    public void setUnicodeName(final byte[] unicodeName) {
        if (unicodeName != null) {
            this.unicodeName = new byte[unicodeName.length];
            System.arraycopy(unicodeName, 0, this.unicodeName, 0,
                             unicodeName.length);
        } else {
            this.unicodeName = null;
        }
        data = null;
    }
}

File: readium/shared/src/commonMain/kotlin/org/readium/r2/shared/util/zip/compress/archivers/zip/AbstractUnicodeExtraField.kt
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Ported to Kotlin Multiplatform for Readium from the vendored Commons Compress subset
// (Java sources: `AbstractUnicodeExtraField.java`, `UnicodePathExtraField.java` and
// `UnicodeCommentExtraField.java`, phase 05b).

package org.readium.r2.shared.util.zip.compress.archivers.zip

import org.readium.r2.shared.util.zip.compress.utils.Crc32

/**
 * A common base class for Unicode extra information extra fields.
 */
internal abstract class AbstractUnicodeExtraField : ZipExtraField {

    /** The CRC32 checksum of the file name or comment as encoded in the central directory. */
    var nameCRC32: Long = 0
        private set

    private var _unicodeName: ByteArray? = null

    private var data: ByteArray? = null

    protected constructor()

    /**
     * Assembles as unicode extension from the name/comment and encoding of the original zip entry.
     *
     * @param text the file name or comment.
     * @param bytes the bytes actually written to the archive.
     */
    protected constructor(text: String, bytes: ByteArray, off: Int = 0, len: Int = bytes.size) {
        val crc32 = Crc32()
        crc32.update(bytes, off, len)
        nameCRC32 = crc32.value
        _unicodeName = text.encodeToByteArray()
    }

    /** The UTF-8 encoded name. */
    val unicodeName: ByteArray?
        get() = _unicodeName?.copyOf()

    private fun assembleData(): ByteArray? {
        val unicodeName = _unicodeName ?: return null
        val data = ByteArray(5 + unicodeName.size)
        // version 1
        data[0] = 0x01
        ZipLong.getBytes(nameCRC32).copyInto(data, 1)
        unicodeName.copyInto(data, 5)
        this.data = data
        return data
    }

    override val centralDirectoryData: ByteArray?
        get() = (data ?: assembleData())?.copyOf()

    override val centralDirectoryLength: ZipShort
        get() = ZipShort((data ?: assembleData())?.size ?: 0)

    override val localFileDataData: ByteArray?
        get() = centralDirectoryData

    override val localFileDataLength: ZipShort
        get() = centralDirectoryLength

    override fun parseFromCentralDirectoryData(buffer: ByteArray, offset: Int, length: Int) {
        parseFromLocalFileData(buffer, offset, length)
    }

    override fun parseFromLocalFileData(buffer: ByteArray, offset: Int, length: Int) {
        if (length < 5) {
            throw ZipException("UniCode path extra data must have at least 5 bytes.")
        }

        val version = buffer[offset].toInt()
        if (version != 0x01) {
            throw ZipException("Unsupported version [$version] for UniCode path extra data.")
        }

        nameCRC32 = ZipLong.getValue(buffer, offset + 1)
        _unicodeName = buffer.copyOfRange(offset + 5, offset + length)
        data = null
    }
}

/**
 * Info-ZIP Unicode Path Extra Field (0x7075).
 *
 * Stores the UTF-8 version of the file name field as stored in the local header and central
 * directory header.
 */
internal class UnicodePathExtraField : AbstractUnicodeExtraField {

    constructor() : super()

    /**
     * Assembles as unicode path extension from the name given as text as well as the encoded
     * bytes actually written to the archive.
     */
    constructor(name: String, bytes: ByteArray, off: Int = 0, len: Int = bytes.size) :
        super(name, bytes, off, len)

    override val headerId: ZipShort
        get() = UPATH_ID

    companion object {
        val UPATH_ID: ZipShort = ZipShort(0x7075)
    }
}

/**
 * Info-ZIP Unicode Comment Extra Field (0x6375).
 *
 * Stores the UTF-8 version of the file comment as stored in the central directory header.
 */
internal class UnicodeCommentExtraField : AbstractUnicodeExtraField {

    constructor() : super()

    /**
     * Assembles as unicode comment extension from the comment given as text as well as the
     * encoded bytes actually written to the archive.
     */
    constructor(comment: String, bytes: ByteArray, off: Int = 0, len: Int = bytes.size) :
        super(comment, bytes, off, len)

    override val headerId: ZipShort
        get() = UCOM_ID

    companion object {
        val UCOM_ID: ZipShort = ZipShort(0x6375)
    }
}
